src/Confluent.Kafka/Producer.cs

```diff
@@ -675,14 +676,29 @@ public int Flush()
         /// </remarks>
         public void Dispose()
         {
+            // TODO: If this method is called in a finalizer, can callbackTask potentially be null?
             topicHandles.Dispose();
 
             if (!this.manualPoll)
             {
-                // Note: It's necessary to wait on callbackTask before disposing kafkaHandle.
-                // TODO: If called in a finalizer, can callbackTask be potentially null?
                 callbackCts.Cancel();
-                callbackTask.Wait();
+                try
+                {
+                    // Note: It's necessary to wait on callbackTask before disposing kafkaHandle
+                    // since the poll loop makes use of this.
+                    callbackTask.Wait();
+                }
+                catch (AggregateException e)
```

Comment:
Why is an exception better than waiting for the task to exit?

Reply:
The old code was intermittently throwing the AggregationException (InnerException == TaskCanceledException). I don't understand where the race condition is (i did some testing to try and find it), and as far as I can tell the old code is valid. 

This definitely fixes the problem, and is valid, and I propose not spending any more time on it :-).

Reply:
it's not